feat: support sigv4 signing

obanby opened the pull request:
- Added Tripper to enable signing requests
- Added Signer to implement the AWS sigv4 signing standard

I based my PR on this comment https://github.com/grafana/xk6-output-prometheus-remote/pull/161#issuecomment-1934440630 and  https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html

Note: I don't have deep knowledge about the sigv4, so please provide as much feedback as possible, and I will be happy to address it :) 

Addresses: https://github.com/grafana/xk6-output-prometheus-remote/issues/93

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/xk6-output-prometheus-remote?pullRequest=169) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/xk6-output-prometheus-remote?pullRequest=169) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/xk6-output-prometheus-remote?pullRequest=169) it.</sub>

mike-sol commented:
This is great, enables easy export to Amazon Managed Prometheus. 👍

mike-sol commented:
Can I buy someone a coffee / beer / etc to get this reviewed and merged?

olegbespalov commented:
Hey @mike-sol !

The pull request is on our radar, but the team is busy with the upcoming K6 release. We'll return to it once we finish. Sorry for the delay!

olegbespalov commented on pkg/sigv4/sigv4.go:
```diff
@@ -0,0 +1,200 @@
+
+	d.payloadHash = payloadHash
+	d.addRequiredHeaders(req)
+	d.canonicalHeaders, d.signedHeaders = d.getCanonicalAndSignedHeaders(req)
```
Is there a big reason to keep the state (`canonicalHeaders` and `signedHeaders` and request related values) on the singer level instead of passing them downstream?

obanby replied:
They were used in multiple places and I didn't want to pass them individually as arguments. 

I changed it as per your suggestion.

olegbespalov commented on pkg/sigv4/sigv4.go:
```diff
@@ -0,0 +1,200 @@
+	var headers []string
+	var signedHeaders []string
+
+	for key, value := range req.Header {
```
I'm checking the [AWS SDK implementation](https://github.com/aws/aws-sdk-go-v2/blob/main/aws/signer/v4/v4.go#L434-L436), and it has a bit more complicated logic, like skipping from signing some keys, do you think we need something similar here?

The same probably applies to the query params

obanby replied:
addressed in [f3e0598](https://github.com/grafana/xk6-output-prometheus-remote/pull/169/commits/f3e0598fb8897a0600f8f5030b8696c25e1553d0)

olegbespalov commented on pkg/sigv4/tripper_test.go:
```diff
@@ -0,0 +1,35 @@
+)
+
+func TestTripper_request_includes_required_headers(t *testing.T) {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Can we make this test `t.Parallel()` ?

obanby replied:
addressed in [f3e0598](https://github.com/grafana/xk6-output-prometheus-remote/pull/169/commits/f3e0598fb8897a0600f8f5030b8696c25e1553d0)

olegbespalov commented on pkg/sigv4/sigv4.go:
```diff
@@ -0,0 +1,200 @@
+
+func (d *DefaultSigner) getPayloadHash(req *http.Request) (string, error) {
+	if req.Body == nil {
+		return hex.EncodeToString(sha256.New().Sum(nil)), nil
```
It could be just a [const](https://github.com/aws/aws-sdk-go-v2/blob/main/aws/signer/internal/v4/const.go#L6), like it's done here.

obanby replied:
addressed in [f3e0598](https://github.com/grafana/xk6-output-prometheus-remote/pull/169/commits/f3e0598fb8897a0600f8f5030b8696c25e1553d0)

olegbespalov reviewed (commented):
Hi @obanby!

Thanks for the PR, and sorry for the delay in reviewing!

First of all, I'd like to say that, like you, I don't have a deep knowledge of sigv4. That's why I checked the AWS Go SDK to familiarize myself with it and made some of my comments based on the logic that I found implemented there.

Another question is, have you tested these changes if they work?

obanby commented:
> Hi @obanby!
> 
> Thanks for the PR, and sorry for the delay in reviewing!
> 
> First of all, I'd like to say that, like you, I don't have a deep knowledge of sigv4. That's why I checked the AWS Go SDK to familiarize myself with it and made some of my comments based on the logic that I found implemented there.
> 
> Another question is, have you tested these changes if they work?

Thanks @olegbespalov, yes I have tested with our aws managed prometheus.

obanby commented on pkg/sigv4/sigv4.go:
```diff
@@ -0,0 +1,200 @@
+	Sign(req *http.Request) error
+}
+
+type DefaultSigner struct {
```
I can do that!

Just for my own knowledge, will maintainability of this be easier due to potential to reuse, or are there any other reasons?

obanby commented on pkg/sigv4/tripper.go:
```diff
@@ -0,0 +1,35 @@
+	AwsAccessKeyID     string
+}
+
+func NewRoundTripper(config *Config, next http.RoundTripper) *Tripper {
```
addressed in [f3e0598](https://github.com/grafana/xk6-output-prometheus-remote/pull/169/commits/f3e0598fb8897a0600f8f5030b8696c25e1553d0)

olegbespalov commented on pkg/sigv4/utils.go:
```diff
@@ -0,0 +1,98 @@
+
+// stripExcessSpaces will rewrite the passed in slice's string values to not
+// contain multiple side-by-side spaces.
+func stripExcessSpaces(str string) string {
```
This method (and maybe others from this file) looks utterly identical to the one from AWS SDK. Unfortunately, I'm not sure if we can accept it as it is because of licensing. Its performance is also less of a concern here. Can we rewrite it (and other methods if needed) by hand?

obanby replied:
Yes these ones where a ported straight for the sdk. I can try to rewrite it by hand. I will look into the test cases for it too and see If I can get them to pass with the new rewrite.

obanby replied:
There are couple more here that ported and made slight changes, would you like me to look into rewriting them?

olegbespalov replied:
I mean, the rule of thumb is that we should not copy code directly, so yeah, if rewriting it on your own is preferable. But since, in any case, this PR is aiming to re-implement some logic to make the output AWS compatible, which still forces us that or another way use similar approaches so I believe at some point, we should still left some credits to AWS SDK with something like `logic is highly inspired by the AWS SDK` or any better referencing

obanby replied:
Yes thats exactly it, I initially didn't want to rewrite it since I wanted the exact aws implementation. My thinking is we want to use the SDK behaviour, but without including it, so I figured porting it would be less risk of mistakes.

So are you okay if I add a comment with attributions instead of rolling a new one?

olegbespalov replied:
No, that's not what I meant, sorry :cry: I was saying that in all cases, we need to mention that logic of that package tries to mimic AWS SDK. But having an utterly identical method, even if we reference to AWS SDK, isn't way to go

olegbespalov commented on pkg/sigv4/sigv4.go:
```diff
@@ -0,0 +1,254 @@
+	// initialize noEscape array. This way we can avoid using init() functions
+	for i := 0; i < len(noEscape); i++ {
+		// AWS expects every character except these to be escaped
+		noEscape[i] = (i >= 'A' && i <= 'Z') ||
```
If we populate this in a constructor, let's move this to a `Signer` property (since it's a state), and if needed, just pass downstream any method. The rationale is to avoid using a global state.

joanlopez reviewed (commented):
Hey @obanby, thanks for your contribution! 🙌🏻 

I have reviewed this PR as well, and I don't have any other comments beyond what @olegbespalov has already mentioned. So, feel free to try to address them all and I'll be happy to give my 👍🏻 once ready!

obanby commented:
Thanks @olegbespalov @joanlopez, I addressed all the comments. 

One thing I noticed, but didn't know how to go about it is: erroring out if only partial config is passed in. Say a user sets the the SIGV4_REGION without setting the access key. Any ideas how to fail with an informative message?

olegbespalov commented:
Hey @obanby 

Thanks for the update; I'll go through it at the beginning of next week!

> One thing I noticed, but didn't know how to go about it is: erroring out if only partial config is passed in. Say a user sets the the SIGV4_REGION without setting the access key. Any ideas how to fail with an informative message?

You could try collecting the validation errors in the slice, something like:

```go
errs := make([]string, 0)

if len(strings.TrimSpace(config.Region)) == 0 {
	errs = append(errs, "sigV4 config `Region` must be set")
}

if len(strings.TrimSpace(config.AwsSecretAccessKey)) == 0 {
	errs = append(errs, "sigV4 config `AwsSecretAccessKey` must be set")
}

if len(strings.TrimSpace(config.AwsAccessKeyID)) == 0 {
	errs = append(errs, "sigV4 config `AwsAccessKeyID` must be set")
}

if len(errs) > 0 {
	return nil, errors.New("can't initialize a sigv4 round tripper:" + strings.Join(errs, ", "))
}
```

obanby commented:
Thanks @olegbespalov, but to clarify. The current error handling should have caught that `config.AwsSecretAccessKey` is not set, and returns the error to the tripper client. However, that error doesn't correctly get propagated to console. So the user end up with a 403 unauthorized, because they didn't see any errors about sigv4. So my question is, how can this error propagate all the way up (preferably also short circuit the test)

olegbespalov commented:
Right, then I believe we should add validation somewhere around `RemoteConfig` e.g.  https://github.com/grafana/xk6-output-prometheus-remote/pull/169/files#diff-35f036a9d95de5edce2e31d56e65f748e31a64950ca6af842f241bf8dc01bf7fR126

```
sigvcfg := &sigv4.Config{
	Region:             conf.SigV4Region.String,
	AwsAccessKeyID:     conf.SigV4AccessKey.String,
	AwsSecretAccessKey: conf.SigV4SecretKey.String,
}

err := sigvcfg.Validate()
if err != nil {
    return nil, err
}

hc.SigV4 = sigvcfg
```

obanby commented:
Sounds good, I will try that. 

I will leave it till your review next week and address it then with any other comments you have :)

olegbespalov commented on pkg/sigv4/sigv4.go:
```diff
@@ -0,0 +1,244 @@
+	"time"
+)
+
+type Signer interface {
```
If I'm not mistaken, we could make this interface (and some other structs in this package) unexported, which explicitly marks them for internal only (package) usage, helping with possible refactoring in the future and protecting other usage

olegbespalov commented:
Hey @obanby !

Just in case, I've reviewed the rest of the changes, and it looks good; maybe the only missing part is [the validation that we discussed before](https://github.com/grafana/xk6-output-prometheus-remote/pull/169#issuecomment-2407775184) to advise customers faster about misconfiguration.

So, thanks for implementing that!

obanby commented:
Thanks @olegbespalov was just taking sometime to see how to go about it. My last commit I addressed the partial commit issue, that's kinda what I was pointing out, see if you like that and we can go from there :)

olegbespalov commented on pkg/remotewrite/config.go:
```diff
@@ -110,6 +124,22 @@ func (conf Config) RemoteConfig() (*remote.HTTPConfig, error) {
 		hc.TLSConfig.Certificates = []tls.Certificate{cert}
 	}
 
+	if isSigV4PartiallyConfigured(conf.SigV4Region, conf.SigV4AccessKey, conf.SigV4SecretKey) {
+		return nil, errors.New(
+			"sigv4 seems to be partially configured. All of " +
+				"K6_PROMETHEUS_RW_SIGV4_REGION, K6_PROMETHEUS_RW_SIGV4_ACCESS_KEY, K6_PROMETHEUS_RW_SIGV4_SECRET_KEY " +
+				"must all be set. Unset all to bypass sigv4",
+		)
+	}
+
+	if conf.SigV4Region.Valid && conf.SigV4AccessKey.Valid && conf.SigV4SecretKey.Valid {
+		hc.SigV4 = &sigv4.Config{
+			Region:             conf.SigV4Region.String,
+			AwsAccessKeyID:     conf.SigV4AccessKey.String,
+			AwsSecretAccessKey: conf.SigV4SecretKey.String,
+		}
+	}
```
`isSigV4PartiallyConfigured` seems like a functional copy of `sigv4.Config.validate()`, it has a minor UX benefit that it explicitly refers to environment variables, but still, to be honest, it's better to have a single version so I have a minor suggestion to get rid of it and just make `sigv4.Config.validate()` exported and execute here, something like:

```suggestion
	if conf.SigV4Region.Valid && conf.SigV4AccessKey.Valid && conf.SigV4SecretKey.Valid {
		hc.SigV4 = &sigv4.Config{
			Region:             conf.SigV4Region.String,
			AwsAccessKeyID:     conf.SigV4AccessKey.String,
			AwsSecretAccessKey: conf.SigV4SecretKey.String,
		}
		
		if err := hc.SigV4.Validate(); err != nil {
		    return nil, fmt.Errorf(
			"sigv4 seems to be partially configured. All of " +
				"K6_PROMETHEUS_RW_SIGV4_REGION, K6_PROMETHEUS_RW_SIGV4_ACCESS_KEY, K6_PROMETHEUS_RW_SIGV4_SECRET_KEY " +
				"must all be set: %w.", err),
		}
	}
```

How does that sound?

obanby replied:
The observation I have is that user can set `K6_PROMETHEUS_RW_SIGV4_ACCESS_KEY` and `K6_PROMETHEUS_RW_SIGV4_ACCESS_KEY` but forgot to set the region env variable. Now they will get 403 error but the error will be coming from upstream and won't be caught early since this condition will resolve to false ` conf.SigV4Region.Valid && conf.SigV4AccessKey.Valid && conf.SigV4SecretKey.Valid`

olegbespalov replied:
Good point! :+1:

olegbespalov reviewed (approved):
Looks good to me! :rocket: Thanks for the contribution! :bow: 

Let's wait till @joanlopez will give a final look/feedback!

obanby commented:
I addressed some of the issues flagged by the CI lint job. 

@olegbespalov @joanlopez please have a look. I hope we can get this in soon :)

obanby commented:
@olegbespalov @joanlopez, when will we be able to merge this please?

olegbespalov commented:
Hey @obanby, you have my approval, but we still need the review from @joanlopez for merging. 

In the meantime, could you please address comments from the linter? Most of them look like auto-fixes, so it's just a matter of running `golangci-lint run --fix` against the files. You could also validate if that helps by running the `make lint` command locally.

obanby commented:
Thanks all, I have address the changes flagged by `make lint`.

There are two missing, one is not related to my changes, and the other will require adding another function to breeak down the method, but doesn't seem like the approach to take. 
```
pkg/remotewrite/config.go:289    funlen  Function 'parseEnvs' is too long (84 > 80)
pkg/remotewrite/trend.go:117:21  gosec   G115: integer overflow conversion int -> uint16
```

olegbespalov commented:
@obanby, no worries. We will take care of these linter issues. We'll merge this PR soon. Thanks again for your contribution! :bow: :tada:

joanlopez commented:
Thank you @obanby for your contribution! 🙇🏻 
Thanks @olegbespalov for driving this to getting it merged! 💟
